opr/grading.go

```diff
@@ -50,79 +52,64 @@ func CalculateGrade(avg [20]float64, opr *OraclePriceRecord) float64 {
-	// miner's identity, because if you do, you have to beat that miner to get any reward.
-	// If you don't use some other miner's identity, you only have to place in the top 10
-	// to be rewarded.
-	IDs := make(map[string]*OraclePriceRecord)
-	var nlist []*OraclePriceRecord
+	// Make sure we have the difficulty calculated for all items in the list.
 	for _, v := range list {
-		id := factom.ChainIDFromStrings(v.FactomDigitalID)
-		last := IDs[id]
-		if last != nil {
-			if v.Difficulty < last.Difficulty {
-				continue
-			}
-		}
-		IDs[id] = v
-		nlist = append(nlist, v)
-	}
-	list = nlist
-
-	last := len(list)
-	// Throw away all the entries but the top 50 in difficulty
-	// bubble sort because I am lazy.  Could be replaced with about anything
-	for j := 0; j < len(list)-1; j++ {
-		for k := 0; k < len(list)-j-1; k++ {
-			d1 := list[k].Difficulty
-			d2 := list[k+1].Difficulty
-			if d1 == 0 || d2 == 0 {
-				//panic("Should not be here")
-			}
-			if d1 < d2 { // sort the smallest difficulty to the end of the list
-				list[k], list[k+1] = list[k+1], list[k]
-			}
-		}
+		v.Difficulty = v.ComputeDifficulty(v.Entry.ExtIDs[0])
 	}
+
+	// Throw away all the entries but the top 50 on pure difficulty alone.
+	// Note that we are sorting in descending order.
+	sort.SliceStable(list, func(i, j int) bool { return list[i].Difficulty > list[j].Difficulty })
+
 	if len(list) > 50 {
-		last = 50
+		list = list[:50]
 	}
-	// Go through and throw away entries that are outside the average or on a tie, have the worst difficulty
-	// until we are only left with 10 entries to reward
-	for i := last; i >= 10; i-- {
+	for i := len(list); i >= 10; i-- {
 		avg := Avg(list[:i])
 		for j := 0; j < i; j++ {
 			CalculateGrade(avg, list[j])
 		}
-		// bubble sort the worst grade to the end of the list. Note that this is nearly sorted data, so
-		// a bubble sort with a short circuit is pretty darn good sort.
-		for j := 0; j < i-1; j++ {
-			cont := false                // If we can get through a pass with no swaps, we are done.
-			for k := 0; k < i-j-1; k++ { // yes, yes I know we can get 2 or 3 x better speed playing with indexes
-				if list[k].Grade > list[k+1].Grade { // bit it is tricky.  This is good enough.
-					list[k], list[k+1] = list[k+1], list[k] // sort first by the grade.
-					cont = true                             // any swap means we continue to loop
-				} else if list[k].Grade == list[k+1].Grade { // break ties with PoW.  Where data is being shared
-					if list[k].Difficulty < list[k+1].Difficulty { // we will have ties.
-						//list[k], list[k+1] = list[k+1], list[k]
-						cont = true // any swap means we continue to loop
-					}
-				}
-			}
-			if !cont { // If we made a pass without any swaps, we are done.
-				break
-			}
-		}
+		// Because this process can scramble the sorted fields, we have to resort with each pass.
+		sort.SliceStable(list[:i], func(i, j int) bool { return list[i].Difficulty > list[j].Difficulty })
```

Comment:
Instead of sorting twice it should be doable in one go: https://stackoverflow.com/a/36122804

Reply:
i used this in my unit test:
https://github.com/pegnet/pegnet/blob/4b9d0d5f72dcdf5105415560519d7007d564ae29/opr/grading_test.go#L101-L108

Reply:
A bit shorter, but a bit less readable (I am big on readability :D). One or the other is fine to me.